docs/docs/tutorial-rag.md

```diff
@@ -0,0 +1,287 @@
+        from openai import OpenAI
+        context = get_most_relevant_document(question)
+        client = OpenAI()
+        query = f"""Use the following information to answer the subsequent question. If the answer cannot be found, write "I don't know."
+        Context:
+        \"\"\"
+        {context}
+        \"\"\"
+        Question: {question}"""
+        response = client.chat.completions.create(
+            model=self.model_name,
+            messages=[
+                {"role": "system", "content": self.system_message},
+                {"role": "user", "content": query},
+            ],
+            temperature=0.0,
+            response_format={"type": "text"},
+        )
+        answer = response.choices[0].message.content
+        return {'answer': answer, 'context': context}
+
+weave.init('rag-qa')
+model = RAGModel(
+    system_message="You are an expert in finance and answer questions related to finance, financial services, and financial markets. When responding based on provided information, be sure to cite the source."
+)
+
+@weave.op()
+async def context_precision_score(question, prediction):
+    context_precision_prompt = """Given question, answer and context verify if the context was useful in arriving at the given answer. Give verdict as "1" if useful and "0" if not with json output. 
+    Output in only valid JSON format.
+
+    question: {question}
+    context: {context}
+    answer: {answer}
+    verification: """
+    client = OpenAI()
+
+    prompt = context_precision_prompt.format(
+        question=question,
+        context=prediction['context'],
+        answer=prediction['answer'],
+    )
+
+    response = client.chat.completions.create(
+        model="gpt-4-turbo-preview",
+        messages=[{"role": "user", "content": prompt}],
+        response_format={ "type": "json_object" }
+    )
+    response_message = response.choices[0].message
+    response = json.loads(response_message.content)
+    return {
+        "verdict": int(response["verdict"]) == 1,
+    }
+
+questions = [
+    {"question": "What significant result was reported about Zealand Pharma's obesity trial?"},
+    {"question": "How much did Berkshire Hathaway's cash levels increase in the fourth quarter?"},
+    {"question": "What is the goal of Highmark Health's integration of Google Cloud and Epic Systems technology?"},
+    {"question": "What were Rivian and Lucid's vehicle production forecasts for 2024?"},
+    {"question": "Why was the Norwegian Dawn cruise ship denied access to Mauritius?"},
+    {"question": "Which company achieved the first U.S. moon landing since 1972?"},
+    {"question": "What issue did Intuitive Machines' lunar lander encounter upon landing on the moon?"}
+]
+
+evaluation = weave.Evaluation(dataset=questions, scorers=[context_precision_score])
+import asyncio 
+asyncio.run(evaluation.evaluate(model))
+```
+
+# Conclusion
```

Comment:
@scottire here as "What's next" we could point them to a full repo to clone and start with -- especially for beginners having this central config file to just pop in HF models and use pre-defined metrics could be interesting. I could change mine to not have LC for example (still working on making it simpler and easier)?